daemon/images.go

```diff
@@ -113,3 +75,37 @@ func (d *Daemon) getAllowedAutomatedResources(ctx context.Context) (resources, e
 	}
 	return result, nil
 }
+
+func calculateChanges(logger log.Logger, workloads []cluster.Workload, imageRepos update.ImageRepos) *update.Automated {
+	changes := &update.Automated{}
+
+	for _, workload := range workloads {
+	containers:
+		for _, container := range workload.ContainersOrNil() {
+			currentImageID := container.Image
+			pattern := policy.GetTagPattern(workload.Policies, container.Name)
```

Comment:
Before I factored this out the policies were collected from the `candidateWorkloads`, which would have required me to include the `candidateWorkloads` as a parameter to this function and did not look right to me as it collected information from two sources.

Removing it broke the semver test but that was due to how the daemon mock is constructed, fixed that by adding the `policy.Set{}` to the workload mock.

If there was a specific reason we collected the policies from the candidates and not the workload itself, please tell and I will revert the change.

Reply:
Let's see, so the previous formulation was

```
1. find all the manifests in the repo that might qualify for an update
2. find the resources in the cluster that correspond to those
3. calculate which of the latter can be updated, looking at the policies from the manifests
```

the new formulation is

```
1. find all the manifests in the repo that might qualify for an update
2. find the resources in the cluster that correspond to those
3. calculate which of the latter can be updated, looking at the policies they have
```

The difference is that the resources in the cluster might not be up to date with respect to the policies -- if they have not been synced yet, for instance.

It's difficult to argue that the original formulation is entirely correct. On balance, though, it's better to preserve the behaviour when it's not clear a change improves on it.

Reply:
Looking again, the existing behaviour is pretty weird. It looks at the image from the cluster, but ultimately it will be making changes to the manifests. Empirically, this mostly works fine -- the problem is more that it's hard to reason about, therefore hard to change :thinking:

Anyway I don't think we need to address that right now; the simplest change would just be to include the manifests (`candidateWorkloads`) in the calculation.

Reply:
> On balance, though, it's better to preserve the behaviour when it's not clear a change improves on it.

I am not sure if I agree with you that the change does not improve behaviour.

We now rely on one source for decision making, which is _applied state_ and not _desired state_. As we are in the process of possibly making modifications these should, in my opinion, be made based upon what we are looking at now and not what we may be looking at in the future.

Reply:
> We now rely on one source for decision making

`workloads` is obtained by asking the cluster for all the workloads that are marked as automated _in the git repo_. So as it stands, this code decides what qualifies for an update (the automation annotation) based on the file in the repo, and decides how to update them (the tag filter annotation) based on the cluster.

Reply:
Gotcha, would it make sense to change this to collect automated workloads from the cluster instead (in a new PR) to ensure we are only making modifications to things the user is actually running?

Reply:
If I had to chose a direction to take things, I would say treat the git repo as canonical and refer to the cluster to determine if operations are (likely to be) safe. Does that make sense from your point of view?